src/main/java/com/epam/grid/engine/provider/utils/DirectoryPathUtils.java

```diff
@@ -0,0 +1,111 @@
+        final StringBuilder properDir = new StringBuilder();
+        final File adjustableFolderPath = new File(adjustableFolder);
+        if (adjustableFolderPath.isAbsolute()) {
+            if (!adjustableFolder.startsWith(gridEngineFolder)) {
+                final String errorMessage = "Provided path is absolute, but doesn't begin with Grid Engine folder";
+                log.error(errorMessage);
+                throw new IllegalStateException(errorMessage);
+            }
+            if (!adjustableFolderPath.exists()) {
+                try {
+                    if (adjustableFolderPath.mkdirs()) {
+                        log.info("Directory with path " + adjustableFolderPath + " was created.");
+                        grantAllPermissionsToFolder(adjustableFolderPath);
+                    } else {
+                        final String errorMessage = "Failed to create directory with path " + adjustableFolderPath;
+                        log.error(errorMessage);
+                        throw new IOException(errorMessage);
+                    }
+                } catch (final Exception exception) {
+                    log.error(exception.getMessage());
+                    exception.printStackTrace();
+                }
+            }
+            properDir.append(adjustableFolderPath);
+        } else {
+            if (gridEngineFolder.endsWith(FORWARDSLASH)) {
+                properDir.append(gridEngineFolder).append(adjustableFolderPath);
+            } else {
+                properDir.append(gridEngineFolder).append(FORWARDSLASH).append(adjustableFolderPath);
+            }
+            log.info("Provided directory was changed to " + properDir);
+            final File gridEnginePath = new File(properDir.toString());
+            if (!gridEnginePath.exists()) {
+                try {
+                    if (gridEnginePath.mkdirs()) {
+                        log.info(gridEnginePath + " Grid Engine folder was created.");
+                        grantAllPermissionsToFolder(gridEnginePath);
+                    } else {
+                        final String errorMessage = "Failed to create directory with path " + gridEnginePath;
+                        log.error(errorMessage);
+                        throw new IOException(errorMessage);
+                    }
+                } catch (final Exception exception) {
+                    log.error(exception.getMessage());
+                    exception.printStackTrace();
+                }
+            }
+        }
+        return properDir.toString();
+    }
+
+    private static void grantAllPermissionsToFolder(final File directory) {
+        try {
+            final Set<PosixFilePermission> permissions = PosixFilePermissions.fromString(ALL_PERMISSIONS_STRING);
+            Files.setPosixFilePermissions(directory.toPath(), permissions);
+        } catch (final IOException exception) {
```

Comment:
fixed

Reply:
So now, we catch a checked exception and somehow handle it (by throw an unchecked one), good, but actually I would handle IOException not here but in higher level ( in `buildProperDir`).
Main rule for such cases is that:
 - low level methods like `grantAllPermissionsToFolder` shouldn't decide what to do with exceptional situation but rather they should just inform about it (so throws a checked exception)
 - method with higher level should deal with handling checked exception, because they "know" business logic and "understand" what they need to do (throw unchecked exception and fail or for example retry)

Reply:
done